cloudevents/generic/baseclient.go

```diff
@@ -0,0 +1,209 @@
+			}
+
+			select {
+			case <-ctx.Done():
+				return
+			case err, ok := <-c.cloudEventsOptions.ErrorChan():
+				if !ok {
+					// error channel is closed, do nothing
+					return
+				}
+
+				runtime.HandleError(fmt.Errorf("the cloudevents client is disconnected, %v", err))
+
+				// the cloudevents client network connection is closed, send the receiver stop signal, set the current
+				// client to nil and retry
+				c.sendReceiverSignal(stopReceiverSignal)
+
+				cloudEventsClient = nil
+				c.resetClient(cloudEventsClient)
+
+				<-connBackoffManager.Backoff().C()
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (c *baseClient) publish(ctx context.Context, evt cloudevents.Event) error {
+	now := time.Now()
+
+	if err := c.cloudEventsRateLimiter.Wait(ctx); err != nil {
+		return fmt.Errorf("client rate limiter Wait returned an error: %w", err)
+	}
+
+	latency := time.Since(now)
+	if latency > longThrottleLatency {
+		klog.Warningf(fmt.Sprintf("Waited for %v due to client-side throttling, not priority and fairness, request: %s",
+			latency, evt))
+	}
+
+	sendingCtx, err := c.cloudEventsOptions.WithContext(ctx, evt.Context)
+	if err != nil {
+		return err
+	}
+
+	klog.V(4).Infof("Sent event: %v\n%s", ctx, evt)
+
+	// make sure the current client is the newest
+	c.RLock()
+	defer c.RUnlock()
+
+	if c.cloudEventsClient == nil {
+		return fmt.Errorf("the cloudevents client is not ready")
+	}
+
+	if result := c.cloudEventsClient.Send(sendingCtx, evt); cloudevents.IsUndelivered(result) {
+		return fmt.Errorf("failed to send event %s, %v", evt, result)
+	}
+
+	return nil
+}
+
+func (c *baseClient) subscribe(ctx context.Context, receive receiveFn) {
+	c.Lock()
+	defer c.Unlock()
+
+	// make sure there is only one subscription go routine starting for one client.
+	if c.receiverChan != nil {
+		klog.Warningf("the subscription has already started")
+		return
+	}
+
+	c.receiverChan = make(chan int)
+
+	// start a go routine to handle cloudevents subscription
+	go func() {
+		receiverCtx, receiverCancel := context.WithCancel(context.TODO())
+		cloudEventsClient := c.cloudEventsClient
+
+		for {
+			if cloudEventsClient != nil {
+				// TODO send a resync request
+
+				go func() {
+					if err := cloudEventsClient.StartReceiver(receiverCtx, func(evt cloudevents.Event) {
+						receive(receiverCtx, evt)
+					}); err != nil {
+						runtime.HandleError(fmt.Errorf("failed to receive cloudevents, %v", err))
+					}
+				}()
+			}
+
+			select {
+			case <-ctx.Done():
+				receiverCancel()
+				close(c.receiverChan)
+				return
+			case signal, ok := <-c.receiverChan:
+				if !ok {
+					// receiver channel is closed, stop the receiver
+					receiverCancel()
+					return
+				}
+
+				switch signal {
+				case restartReceiverSignal:
+					klog.V(4).Infof("restart the cloudevents receiver")
+					// make sure the current client is the newest
+					c.RLock()
+					cloudEventsClient = c.cloudEventsClient
+					c.RUnlock()
```

Comment:
do we still need to lock it here? since it is locked at the beginning of the subscribe function

Reply:
yes, we need it, the beginning lock is for the subscribe function, this lock is in a independent go routine (the subscribe lock will be released after this go routine starts)

Reply:
emm, right, thanks for the explanation.